receiver/otlpreceiver/otlp.go

```diff
@@ -0,0 +1,273 @@
+		if err != nil {
+			return
+		}
+
+	})
+
+	return err
+}
+
+func (r *Receiver) registerMetricsConsumer() error {
+	var err = oterr.ErrAlreadyStarted
+
+	r.startMetricsReceiverOnce.Do(func() {
+		r.metricsReceiver, err = metrics.New(r.instanceName, r.metricsConsumer)
+		if err != nil {
+			return
+		}
+		srv := r.grpcServer()
+		collectortrace.RegisterTraceServiceServer(srv, r.traceReceiver)
+	})
+
+	return err
+}
+
+func (r *Receiver) grpcServer() *grpc.Server {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.serverGRPC == nil {
+		r.serverGRPC = observability.GRPCServerWithObservabilityEnabled(r.grpcServerOptions...)
+	}
+
+	return r.serverGRPC
+}
+
+// Shutdown is a method to turn off receiving.
+func (r *Receiver) Shutdown() error {
+	if err := r.stop(); err != oterr.ErrAlreadyStopped {
+		return err
+	}
+	return nil
+}
+
+// start runs all the receivers/services namely, Trace and Metrics services.
+func (r *Receiver) start(host component.Host) error {
+	hasConsumer := false
+	if r.traceConsumer != nil {
+		hasConsumer = true
+		if err := r.registerTraceConsumer(); err != nil && err != oterr.ErrAlreadyStarted {
+			return err
+		}
+	}
+
+	if r.metricsConsumer != nil {
+		hasConsumer = true
+		if err := r.registerMetricsConsumer(); err != nil && err != oterr.ErrAlreadyStarted {
+			return err
+		}
+	}
+
+	if !hasConsumer {
+		return errors.New("cannot start receiver: no consumers were specified")
+	}
+
+	if err := r.startServer(host); err != nil && err != oterr.ErrAlreadyStarted {
+		return err
+	}
+
+	// At this point we've successfully started all the services/receivers.
+	// Add other start routines here.
+	return nil
+}
+
+// stop stops the underlying gRPC server and all the services running on it.
+func (r *Receiver) stop() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var err = oterr.ErrAlreadyStopped
+	r.stopOnce.Do(func() {
+		err = nil
+
+		if r.serverHTTP != nil {
+			_ = r.serverHTTP.Close()
+		}
+
+		if r.ln != nil {
+			_ = r.ln.Close()
+		}
+
+		r.serverGRPC.Stop()
+	})
+	return err
+}
+
+func (r *Receiver) httpServer() *http.Server {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.serverHTTP == nil {
+		var mux http.Handler = r.gatewayMux
+		if len(r.corsOrigins) > 0 {
+			co := cors.Options{AllowedOrigins: r.corsOrigins}
+			mux = cors.New(co).Handler(mux)
+		}
+		r.serverHTTP = &http.Server{Handler: mux}
+	}
+
+	return r.serverHTTP
+}
+
+func (r *Receiver) startServer(host component.Host) error {
+	err := oterr.ErrAlreadyStarted
+	r.startServerOnce.Do(func() {
+		err = nil
+		// Register the grpc-gateway on the HTTP server mux
+		c := context.Background()
+		opts := []grpc.DialOption{grpc.WithInsecure()}
+		endpoint := r.ln.Addr().String()
+
+		_, ok := r.ln.(*net.UnixListener)
+		if ok {
+			endpoint = "unix:" + endpoint
+		}
+
+		// TODO(nilebox): Compilation is broken due to missing gRPC gateway boilerplate,
+		//  see https://github.com/open-telemetry/opentelemetry-proto/pull/110
```

Comment:
@nilebox  Can we move forward without gRPC gateway? We can add support for http/json later.

Reply:
This is now resolved.